### PR TITLE
elmerfem: 9.0 → unstable-2022-08-30

### PR DESCRIPTION
**Pull request by grindhold**
upstream released important features and fixed bugs i needed
patches for

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Wulfsta** [comment]
Does 9.0 have problems (or is the release cadence really slow)? If not, I would prefer two versions similar to how `super-slicer` and `super-slicer-latest` are packaged.

**Comment by grindhold** [comment]
roughly 1 per year although the software develops. i'd deem it slow.

File: pkgs/applications/science/physics/elmerfem/default.nix [hunk 1]
```diff
@@ -2,13 +2,13 @@
...
   pname = "elmerfem";
-  version = "9.0";
+  version = "unstable-2022-08-30";
 
   src = fetchFromGitHub {
     owner = "elmercsc";
     repo = "elmerfem";
-    rev = "release-${version}";
-    sha256 = "VK7jvu4s5d7k0c39XqY9dEzg/vXtX5Yr/09VcuZVQ9A=";
+    rev = "3eafebb";
```
Comment (superherointj) [inline pkgs/applications/science/physics/elmerfem/default.nix:10]:
```suggestion
    rev = "3eafebb4a0287c435376207e15003bf32ffb11e3";
```

**Comment by superherointj** [comment]
Result of `nixpkgs-review pr 189014` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>elmerfem</li>
  </ul>
</details>

**Comment by superherointj** [comment]
Is there a main application in this package for `meta.mainProgram`? (to run in `nix run .#elmerfem`)

**Comment by grindhold** [comment]
I'd name `bin/ElmerGUI`  for that purpose

**Comment by superherointj** [comment]
> I'd name `bin/ElmerGUI` for that purpose

Then, please add it to meta.
And test it with `nix run .#elmerfem`.

**Comment by grindhold** [comment]
```
nix run .#elmerfem
```
leads to 
`error: experimental Nix feature 'nix-command' is disabled; use '--extra-experimental-features nix-command' to override`

```
nix --extra-experimental-features nix-command run .#elmerfem
```
leads to
`error: experimental Nix feature 'flakes' is disabled; use '--extra-experimental-features flakes' to override`

how am i supposed to do this?

**Comment by superherointj** [comment]
> ```
> nix run .#elmerfem
> ```
> leads to 
> `error: experimental Nix feature 'nix-command' is disabled; use '--extra-experimental-features nix-command' to override`
> 
> ```
> nix --extra-experimental-features nix-command run .#elmerfem
> ```
> leads to
> `error: experimental Nix feature 'flakes' is disabled; use '--extra-experimental-features flakes' to override`
> 
> how am i supposed to do this? and why does it complain about flakes? i'm not using flakes here
> 

Sorry. It is a flakes command.

**Comment by superherointj** [comment]
Result of `nixpkgs-review pr 189014` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>elmerfem</li>
  </ul>
</details>

**Comment by superherointj** [comment]
> I'd name `bin/ElmerGUI` for that purpose

`bin/ElmerGUI` did not run for me. Complains of missing: libSM, libtiff, etc. (Which you can add to the package, but I have not checked how far this rabbit hole goes.)

**Comment by superherointj** [comment]
I have added `mainProgram`. Feel free to pull and squash it (if you can).
Also reorder attributes per phases order (unpack => patch => configure => build => test => install)

**Comment by grindhold** [comment]
done :)

**Comment by superherointj** [comment]
Result of `nixpkgs-review pr 189014` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>elmerfem</li>
  </ul>
</details>

**Comment by superherointj** [comment]
> done :)

Did you check the matter of missing libraries when running `bin/ElmerGUI`?

**Comment by superherointj** [comment]
Also, feel free to pull changes and squash.

File: pkgs/applications/science/physics/elmerfem/default.nix [hunk 2]
```diff
@@ -1,68 +1,78 @@
...
-    rev = "release-${version}";
-    sha256 = "VK7jvu4s5d7k0c39XqY9dEzg/vXtX5Yr/09VcuZVQ9A=";
+    rev = "3eafebb4a0287c435376207e15003bf32ffb11e3";
+    hash = "sha256-2lVJmbFBDnUo+bw/AILzUncOJM4B7eAW0mosmsAWANI=";
   };
 
+  storepath = placeholder "out";
+
+  cmakeFlags = [
+    "-DELMER_INSTALL_LIB_DIR=${storepath}/lib"
```
Comment (SuperSandro2000) [inline pkgs/applications/science/physics/elmerfem/default.nix:32]:
```suggestion
  cmakeFlags = [
    "-DELMER_INSTALL_LIB_DIR=${placeholder "out"}/lib"
```

File: pkgs/applications/science/physics/elmerfem/default.nix [hunk 3]
```diff
@@ -1,68 +1,78 @@
...
+    "-DWITH_OCC:BOOLEAN=TRUE"
+    "-DWITH_VTK:BOOLEAN=TRUE"
+    "-DWITH_ELMERGUI:BOOLEAN=TRUE"
+    "-DCMAKE_INSTALL_LIBDIR=lib"
+    "-DCMAKE_INSTALL_INCLUDEDIR=include"
+    "-DCMAKE_OpenGL_GL_PREFERENCE=GLVND"
+  ];
+
+  preConfigure = ''
+    patchShebangs ./
```
Comment (SuperSandro2000) [inline pkgs/applications/science/physics/elmerfem/default.nix:45]: Can we be more specific?

File: pkgs/applications/science/physics/elmerfem/default.nix [hunk 4]
```diff
@@ -1,68 +1,78 @@
...
+  '';
+
   hardeningDisable = [ "format" ];
```
Comment (SuperSandro2000) [inline pkgs/applications/science/physics/elmerfem/default.nix:48]: Is this still true?

File: pkgs/applications/science/physics/elmerfem/default.nix (hunk 2 above)
Comment (grindhold) [reply pkgs/applications/science/physics/elmerfem/default.nix]: done

File: pkgs/applications/science/physics/elmerfem/default.nix (hunk 3 above)
Comment (grindhold) [reply pkgs/applications/science/physics/elmerfem/default.nix]: apparently not

File: pkgs/applications/science/physics/elmerfem/default.nix (hunk 4 above)
Comment (grindhold) [reply pkgs/applications/science/physics/elmerfem/default.nix:46]: apparently not

**Comment by grindhold** [comment]
every itch should have been scratched, let ofborg do the thing! 

do the thing

File: pkgs/applications/science/physics/elmerfem/default.nix (hunk 3 above)
Comment (SuperSandro2000) [reply pkgs/applications/science/physics/elmerfem/default.nix]: You are sure? Can we at least specific a directory?

Comment (grindhold) [reply pkgs/applications/science/physics/elmerfem/default.nix]: naah. i dropped the whole thing and the build still works. it seems to be unneccessary.

**Pull request closed without merging**